### PR TITLE
Add minimal infrastructure for CI and deployment

### PR DESCRIPTION
**Pull request by sbesson**
- Split the logic into a standalone library and the plugin wrapper for CLI
- Fix integration tests to use omero_cli_upload
- Add top-level GPL license and README
- Add setup.py for deployment to PyPI
- Add .travis.yml with testing logic (omero-test-infra) and PyPI deployment

Proposed tag/release: `0.1.0`.

NB: a next step might be to deprecate/drop the `upload` plugin in core OMERO.py 5.5.0.

### CODE REVIEW

**Comment by joshmoore** [comment]
If it eases immediate testing, no objections to `upload2` but I'm not a big fan. Do you know that the pip installed version won't get picked up? Looking at:

```
        paths = set(self._plugin_paths)
        for x in sys.path:
            x = path(x)
            if x.isdir():
                x = x / "omero" / "plugins"
                if x.exists():
                    paths.add(x)
            else:
                if self.isdebug:
                    print "Can't load %s" % x
        for plugin_path in paths:
            self.loadpath(path(plugin_path))
```
(https://github.com/openmicroscopy/openmicroscopy/blob/1b4872907427db6fa5cf30674bc0f5343dd2847a/components/tools/OmeroPy/src/omero/cli.py#L1500) 

I'd assume last item in `[lib/python + PYTHONPATH]` would win.

**Comment by sbesson** [comment]
Will test in a separate branch.

**Comment by sbesson** [comment]
https://travis-ci.org/sbesson/omero-cli-upload/builds/494820157 suggests indeed the plugin in the `PYTHONPATH` takes precedence

```
+ /opt/omero/server/OMERO.server/bin/omero upload -h
usage: /opt/omero/server/OMERO.server/bin/omero upload [-h] [--test TEST] [-C]
                                                       [-s SERVER] [-p PORT]
                                                       [-g GROUP] [-u USER]
                                                       [-w PASSWORD] [-k KEY]
                                                       [--sudo ADMINUSER] [-q]
                                                       file [file ...]
Upload local files to the OMERO server
Positional Arguments:
  file
Optional Arguments:
  In addition to any higher level options
  -h, --help                        show this help message and exit
  --test TEST                       Test precedence
```

I'll push an extra commit to preserve the plugin name.

File: src/omero_cli_upload.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import path
+import mimetypes
+
+from omero.cli import BaseControl
+
+try:
+    import hashlib
+    hash_sha1 = hashlib.sha1
+except Exception:
+    import sha
```
Comment (manics) [inline src/omero_cli_upload.py:32]: `sha` was replaced by `hashlib` in Python 2.5: https://docs.python.org/2/library/sha.html?#module-sha

**Review by manics — COMMENTED** [review]
The bundled plugin is still being detected first:
```
$ docker run -it --rm --entrypoint bash -u root openmicroscopy/omero-web-standalone

# sed -i 's/client = self.ctx.conn(args)/assert False/'  /opt/omero/web/OMERO.web/lib/python/omero/plugins/upload.py

# /opt/omero/web/venv/bin/pip install https://github.com/ome/omero-cli-upload/archive/dcc71b095409ab3d459531cea9a3a6afbe7678c1.zip

# su - omero-web

$ /opt/omero/web/OMERO.web/bin/omero login <USER>@demo.openmicroscopy.org
Password:
Created session for <USER>@demo.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: 2018-10

$ . /opt/omero/web/venv/bin/activate

$ /opt/omero/web/OMERO.web/bin/omero upload OMERO.py-5.4.10-ice36-b105.zip
Traceback (most recent call last):
  File "/opt/omero/web/OMERO.web/bin/omero", line 130, in <module>
    rv = omero.cli.argv()
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1620, in argv
    cli.invoke(args[1:])
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1095, in invoke
    stop = self.onecmd(line, previous_args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1172, in onecmd
    self.execute(line, previous_args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1254, in execute
    args.func(args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/plugins/upload.py", line 52, in upload
    assert False
AssertionError

$ pip list |grep omero-cli
omero-cli-upload (0.1.0.dev1)
```

This works if I removing the bundled plugin:
```
$ mv OMERO.web/lib/python/omero/plugins/upload.py{,.disable}

$ /opt/omero/web/OMERO.web/bin/omero upload OMERO.py-5.4.10-ice36-b105.zip
Using session for <USER>@demo.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: 2018-10
OriginalFile:200751
```

File: src/omero_cli_upload.py (hunk 1 above)
Comment (sbesson) [reply src/omero_cli_upload.py:32]: :+1: Now we're Python 2.7 only - this should be safe to drop

**Comment by sbesson** [comment]
Thanks for the reproducible scenario in https://github.com/ome/omero-cli-upload/pull/1#pullrequestreview-205136897. I think the problem comes back to the ordering in `sys.path` as discussed https://github.com/ome/omero-cli-upload/pull/1#issuecomment-464657745.

Using the same Docker setup and creating another OMERO.web symlink under `/opt/omero/web/aweb`, I have

```
(venv) -bash-4.2$ /opt/omero/web/OMERO.web/bin/omero upload test
Traceback (most recent call last):
  File "/opt/omero/web/OMERO.web/bin/omero", line 130, in <module>
    rv = omero.cli.argv()
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1620, in argv
    cli.invoke(args[1:])
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1095, in invoke
    stop = self.onecmd(line, previous_args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1172, in onecmd
    self.execute(line, previous_args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/cli.py", line 1254, in execute
    args.func(args)
  File "/opt/omero/web/OMERO.web/lib/python/omero/plugins/upload.py", line 52, in upload
    assert False
AssertionError
(venv) -bash-4.2$ /opt/omero/web/aweb/bin/omero upload test
Server: [localhost:4064]
```

Working around this issue for the lifetime of OMERO 5.4.x via documentation could be an option. that being said, the same problem will happen for any core CLI plugin (not optionally activated via `OMERO_DEV_PLUGINS`) we start decoupling so we might want to agree on the strategy.

**Comment by manics** [comment]
If the only reason to support 5.4 is for the IDR we could remove `upload.py` from `develop` and the `metadata54` branches, and officially release this for 5.5 only?

**Comment by sbesson** [comment]
I opened https://github.com/openmicroscopy/openmicroscopy/pull/5956 which should start the discussion on how to drop/deprecate the upstream `upload` plugin. 

Proposing to merge and release as `0.1.0` so that we can @Manics can start adding features and potentially investigating a package renaming.
Main caveat as highlighted above is that anyone using OMERO 5.4 and the `omero-cli-upload` plugin would have to pay attention to their path order to use the new features.

/cc @manics @joshmoore @jburel

**Comment by joshmoore** [comment]
WFM

File: src/omero_cli_upload.py (hunk 1 above)
Comment (manics) [reply src/omero_cli_upload.py:32]: Do you want to change this for 0.1.0 or leave it for a future cleanup?

Comment (sbesson) [reply src/omero_cli_upload.py:32]: Proposing to leave it. My goal with `0.1.0` was to have a tag with the decoupled plugin matching the code bundled in OMERO.py 5.4.x. Cleanup/additions can then be released in follow-up tags.

**Comment by manics** [comment]
From https://github.com/sbesson/omero-cli-upload/pull/1#issuecomment-467795570
> Interestingly, this feature raises the question of whether `omero-cli-upload` should now be rebranded more generically as `omero-upload` i.e. a standalone Python package containing more than a CLI plugin like `omero-metadata`

If we decide to rename do it before releasing `0.1.0`?

**Comment by sbesson** [comment]
Yes, I'd rather rename the GitHub repository/Python project before the first tag. Which option would you vote for?

**Comment by manics** [comment]
`omero-upload`

**Comment by manics** [comment]
Unless there's potential for confusion with a future omero-uploader client app, then omero-upload-python/omero-python-upload (or something similar) might be better for the github repo name, but keep the pypi package name as omero-upload? I don't mind, `omero-upload` is still fine with me.

**Pull request merged**